runtime/cudaq/platform/default/rest_server/helpers/RestRemoteClient.cpp

```diff
@@ -0,0 +1,223 @@
+                            std::string(), [](const auto &ss, const auto &s) {
+                              return ss.empty() ? s : ss + "," + s;
+                            });
+        if (failed(parsePassPipeline(pipeline, pm, os)))
+          throw std::runtime_error(
+              "Remote rest platform failed to add passes to pipeline (" +
+              errMsg + ").");
+
+        if (failed(pm.run(moduleOp)))
+          throw std::runtime_error(
+              "Remote rest platform: applying IR passes failed.");
+      }
+      std::string mlirCode;
+      llvm::raw_string_ostream outStr(mlirCode);
+      mlir::OpPrintingFlags opf;
+      opf.enableDebugInfo(/*enable=*/true,
+                          /*pretty=*/false);
+      moduleOp.print(outStr, opf);
+      return llvm::encodeBase64(mlirCode);
+    }
+  }
+
+  virtual bool
+  sendRequest(MLIRContext &mlirContext, cudaq::ExecutionContext &io_context,
+              const std::string &backendSimName, const std::string &kernelName,
+              void (*kernelFunc)(void *), void *kernelArgs,
+              std::uint64_t argsSize, std::string *optionalErrorMsg) override {
+    cudaq::RestRequest request(io_context);
+    request.entryPoint = kernelName;
+    if (cudaq::__internal__::isLibraryMode(kernelName)) {
+      request.format = cudaq::CodeFormat::LLVM;
+      if (kernelArgs && argsSize > 0) {
+        cudaq::info("Serialize {} bytes of args.", argsSize);
+        request.args.resize(argsSize);
+        std::memcpy(request.args.data(), kernelArgs, argsSize);
+      }
+
+      if (kernelFunc) {
+        ::Dl_info info;
+        ::dladdr(reinterpret_cast<void *>(kernelFunc), &info);
+        const auto funcName = cudaq::quantum_platform::demangle(info.dli_sname);
+        cudaq::info("RemoteSimulatorQPU: retrieve name '{}' for kernel {}",
+                    funcName, kernelName);
+        request.entryPoint = funcName;
+      }
+    } else {
+      request.passes = serverPasses;
+      request.format = cudaq::CodeFormat::MLIR;
+    }
+
+    request.code = constructKernelPayload(mlirContext, kernelName, kernelFunc,
+                                          kernelArgs, argsSize);
+    if (request.code.empty()) {
+      if (optionalErrorMsg)
+        *optionalErrorMsg =
+            std::string(
+                "Failed to construct/retrieve kernel IR for kernel named ") +
+            kernelName;
+      return false;
+    }
+    request.simulator = backendSimName;
+    request.seed = cudaq::get_random_seed();
+    // Don't let curl adding "Expect: 100-continue" header, which is not
+    // suitable for large requests, e.g., bitcode in the JSON request.
+    //  Ref: https://gms.tf/when-curl-sends-100-continue.html
+    std::map<std::string, std::string> headers{
+        {"Expect:", ""}, {"Content-type", "application/json"}};
+    json requestJson = request;
+    try {
+      static thread_local cudaq::RestClient restClient;
```

Comment:
Why is this `static thread_local`? Is this just an optimization to reduce some initialization?

Reply:
Yea, I thought that since it is a thin wrapper around curl calls and we only need it in this one place, just make it a local static var.

Reply:
> Yea, I thought that since it is a thin wrapper around curl calls and we only need it in this one place, just make it a local static var.

It looks like it only has one `bool` flag in this class, so while I think the code is functional and correct, I'm not sure the optimization (and potential future maintenance complexity) is really worth deviating from standard non-static, stack-based objects in this case. It looks like the same thing applies in `runtime/cudaq/platform/mqpu/helpers/MQPUUtils.cpp`.

Reply:
Fixed in https://github.com/NVIDIA/cuda-quantum/pull/1012/commits/421065ca96ea42fabb18d44c66babd5f6f67af63